pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,316 @@
+			return fmt.Errorf("failed to ensure imagePullSecret: %w", err)
+		}
+		imageName, err := publicDomainForImage(srcContext, sourceImageStreamTag.Image.DockerImageReference)
+		if err != nil {
+			return fmt.Errorf("failed to get the public domain: %w", err)
+		}
+
+		imageStreamImport := &imagev1.ImageStreamImport{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: req.Namespace,
+				Name:      imageStreamName,
+			},
+			Spec: imagev1.ImageStreamImportSpec{
+				Import: true,
+				Images: []imagev1.ImageImportSpec{{
+					From: corev1.ObjectReference{
+						Kind: "DockerImage",
+						Name: imageName,
+					},
+					To: &corev1.LocalObjectReference{Name: imageTag},
+					ReferencePolicy: imagev1.TagReferencePolicy{
+						Type: imagev1.LocalTagReferencePolicy,
+					},
+				}},
+			},
+		}
+
+		r.log.Debugf("creating import %v on cluster %s", imageStreamImport, ctx)
+
+		// ImageStreamImport is not an ordinary api but a virtual one that does the import synchronously
+		if err := client.Create(r.ctx, imageStreamImport); err != nil {
+			return fmt.Errorf("failed to import Image: %w", err)
+		}
+
+		// This should never be needed, but we shouldn't panic if the server screws up
+		if imageStreamImport.Status.Images == nil {
+			imageStreamImport.Status.Images = []imagev1.ImageImportStatus{{}}
+		}
+		if imageStreamImport.Status.Images[0].Image == nil {
+			return fmt.Errorf("imageStreamImport did not succeed: reason: %s, message: %s", imageStreamImport.Status.Images[0].Status.Reason, imageStreamImport.Status.Images[0].Status.Message)
+		}
+
+		log.Debug("Imported successfully")
+	}
+	return nil
+}
+
+func findNewest(isTags map[string]*imagev1.ImageStreamTag) (string, error) {
+	resultContext := ""
+	t := metav1.NewTime(time.Time{})
+	for ctx, isTag := range isTags {
+		if isTag == nil {
+			continue
+		}
+		if t.Before(&isTag.Image.CreationTimestamp) {
+			t = isTag.Image.CreationTimestamp
+			resultContext = ctx
+		}
+	}
+	if resultContext != "" {
+		return resultContext, nil
+	}
+	return "", fmt.Errorf("cannot find the newest isTag from %v", isTags)
+}
+
+const pullSecretName = "registry-cluster-pull-secret"
+
+func (r *reconciler) ensureImagePullSecret(namespace string, client ctrlruntimeclient.Client, log *logrus.Entry) error {
+	secret, mutateFn := r.pullSecret(namespace)
+	return upsertObject(r.ctx, client, secret, mutateFn, log)
+}
+
+// ci-operator uses the release controller configuration to determine
+// the version of OpenShift we create from the ImageStream, so we need
+// to copy the annotation if it exists
+const releaseConfigAnnotation = "release.openshift.io/config"
+
+func imagestream(imageStream *imagev1.ImageStream) (*imagev1.ImageStream, crcontrollerutil.MutateFn) {
+	stream := &imagev1.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: imageStream.Namespace,
+			Name:      imageStream.Name,
+		},
+	}
+	return stream, func() error {
+		if config, set := imageStream.Annotations[releaseConfigAnnotation]; set {
```

Comment:
Can we just copy everything, modulo readOnly fields?

Reply:
Updated.

Thought to do it in another PR.